src/mo_frontend/typing.ml

```diff
@@ -1048,11 +1051,48 @@ and infer_exp'' env exp : T.typ =
       else env
     in
     infer_obj env' obj_sort.it dec_fields exp.at
-  | ObjE exp_fields ->
+  | ObjE (exp_bases, exp_fields) ->
+    let open List in
     check_ids env "object" "field"
-      (List.map (fun (ef : exp_field) -> ef.it.id) exp_fields);
-    let fts = List.map (infer_exp_field env) exp_fields in
-    T.Obj (T.Object, List.sort T.compare_field fts)
+      (map (fun (ef : exp_field) -> ef.it.id) exp_fields);
+    let fts = map (infer_exp_field env) exp_fields in
+    let fls = map (fun {T.lab; _} -> lab) fts in
+    let bases = map (fun b -> infer_exp_promote env b, b) exp_bases in
+
+    (* removing explicit fields from the bases *)
+    let strip (base_t, base) =
+      let s, tfs =
+        try T.as_obj base_t with Invalid_argument _ ->
```

Comment:
I think this is fine and there is no need use as_obj_sub - there is no minimal type to infer is some base has type 'None' and you should just error.

Reply:
I am getting the error (#3405) for `<A <: { a : Nat }>(base : A)`, I don't think we have a `None` in this game...

Reply:
Should be fixed now